wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/urlfactory/URLFactoryBuilder.java

```diff
@@ -0,0 +1,117 @@
+import org.eclipse.che.api.workspace.shared.dto.WorkspaceConfigDto;
+import org.eclipse.che.dto.server.DtoFactory;
+
+/**
+ * Handle the creation of some elements used inside a {@link FactoryDto}.
+ *
+ * @author Florent Benoit
+ */
+@Singleton
+public class URLFactoryBuilder {
+
+  /** Default docker image (if repository has no dockerfile) */
+  protected static final String DEFAULT_DOCKER_IMAGE = "eclipse/ubuntu_jdk8";
+
+  /** Default docker type (if repository has no dockerfile) */
+  protected static final String DEFAULT_MEMORY_LIMIT_BYTES = Long.toString(2000L * 1024L * 1024L);
+
+  protected static final String MACHINE_NAME = "ws-machine";
+
+  private final URLFetcher urlFetcher;
+  private final DevfileManager devfileManager;
+
+  @Inject
+  public URLFactoryBuilder(URLFetcher urlFetcher, DevfileManager devfileManager) {
+    this.urlFetcher = urlFetcher;
+    this.devfileManager = devfileManager;
+  }
+
+  /**
+   * Build a default factory using the provided json file or create default one
+   *
+   * @param jsonFileLocation location of factory json file
+   * @return a factory or null if factory json in not found
+   */
+  public FactoryDto createFactoryFromJson(String jsonFileLocation) {
+    // Check if there is factory json file inside the repository
+    if (jsonFileLocation != null) {
+      final String factoryJsonContent = urlFetcher.fetch(jsonFileLocation);
+      if (!Strings.isNullOrEmpty(factoryJsonContent)) {
+        return DtoFactory.getInstance().createDtoFromJson(factoryJsonContent, FactoryDto.class);
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Build a default factory using the provided devfile
+   *
+   * @param devfileLocation location of devfile
+   * @return a factory or null if devfile is not found
+   */
+  public FactoryDto createFactoryFromDevfile(String devfileLocation) throws BadRequestException {
+    if (devfileLocation != null) {
+      final String devfileYamlContent = urlFetcher.fetch(devfileLocation);
+      if (!Strings.isNullOrEmpty(devfileYamlContent)) {
+        try {
+          WorkspaceConfigImpl wsConfig =
+              devfileManager.validateAndConvert(devfileYamlContent, false);
+          return newDto(FactoryDto.class)
+              .withV(CURRENT_VERSION)
+              .withWorkspace(DtoConverter.asDto(wsConfig));
+        } catch (DevfileFormatException e) {
+          throw new BadRequestException(e.getMessage());
+        } catch (IOException x) {
+          throw new IllegalStateException(x.getLocalizedMessage(), x);
+        }
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Help to generate default workspace configuration
+   *
+   * @param name the name of the workspace
+   * @return a workspace configuration
+   */
+  public WorkspaceConfigDto buildDefaultWorkspaceConfig(String name) {
+
+    Map<String, String> attributes = new HashMap<>();
+    attributes.put(WORKSPACE_TOOLING_EDITOR_ATTRIBUTE, "org.eclipse.che.editor.theia:1.0.0");
```

Comment:
hello, is it possible to grab the latest version from the registry instead of using 1.0.0 ?
the same for machine exec plug-in ?

Reply:
Well, maybe then it should be possible to have the ability to configure workspace to use the latest editor/plugin version instead of grabbing the latest version from the registry in Factory related code?
Do you think it makes sense for workspace configuration as well?

Reply:
I don't remember the ticket issue but I saw discussion around using `latest` instead of `1.0.0` or make version field optional as well

Also, there is still a remaining question, should it be like a che server env var (with default value) the list of all plug-ins set by default ? or to reuse another factory definition or something else.
Let say tomorrow a new plug-in is enabled on some stacks, we need to recompile Che to make the update

Reply:
It definitely must be configured somehow. I think, two properties - editor + plugins list would be enough.

Reply:
@skabashnyuk  it is not used when devfile is present. it is used when there in _no_ devfile :-)
so i prefer to not use word "devfile" here

Reply:
che.factory.default_editor che.factory.default_plugins ?

Reply:
yes